plugins/ida/ida_main.ml

```diff
@@ -137,57 +136,118 @@ let loader path =
             else code, Memmap.add data mem sec) in
   Project.Input.create arch path ~code ~data
 
-let require req check =
-  if check
-  then Ok ()
-  else Or_error.errorf "IDA configuration failure: %s" req
-
 let checked ida_path is_headless =
-  let (/) = Filename.concat in
-  require "path must exist"
-    (Sys.file_exists ida_path) >>= fun () ->
-  require "path must be a folder"
-    (Sys.is_directory ida_path) >>= fun () ->
-  require "can't use headless on windows"
-    (is_headless ==> not Sys.win32) >>= fun () ->
-  require "idaq must exist"
-    (Sys.file_exists (ida_path/"idaq")) >>= fun () ->
-  require "idaq64 must exist"
-    (Sys.file_exists (ida_path/"idaq64")) >>= fun () ->
-  require "idal must exist"
-    (Sys.file_exists (ida_path/"idal")) >>= fun () ->
-  require "idal64 must exist"
-    (Sys.file_exists (ida_path/"idal64")) >>= fun () ->
-  require "bap-ida-python must be installed"
-    (Sys.file_exists
-       (ida_path/"plugins"/"plugin_loader_bap.py"))  >>| fun () ->
-  ida_path
-
+  Bap_ida_check.(check_headless is_headless >>= fun () -> check_path ida_path)
 
 let main () =
   register_source (module Rooter);
   register_source (module Symbolizer);
   register_source (module Reconstructor);
   Project.Input.register_loader name loader
 
-let () =
-  let () = Config.manpage [
-      `S "DESCRIPTION";
-      `P "This plugin provides rooter, symbolizer and reconstuctor services.";
-      `P "If IDA instance is found on the machine, or specified by a
+type 'a p = [ `Explicit of 'a | `auto]
+type headless = bool p
+type mode = [ `m32 | `m64 ] p
+
+let ida_mode is_headless = function
+  | `auto -> None
+  | `Explicit mode -> Some (match mode with
+      | `m32 when is_headless -> `idal
+      | `m64 when is_headless -> `idal64
+      | `m32 -> `idaq
+      | `m64 -> `idaq64)
+
+let bool_of_headless = function
+  | `Explicit x -> x
+  | `auto -> Bap_ida_config.is_headless
+
+let version_of_path path =
+  let path = FilePath.basename path in
+  let re = Str.regexp "[0-9]\\(\\.[0-9]\\)?" in
+  let _x = Str.search_forward re path 0 in
+  try Some (Str.matched_string path)
+  with _ -> None
+
+let find_path paths version =
+  let test path msg = match path with
+    | Some p -> p
+    | None -> eprintf "%s\n" msg; exit 1 in
+  match version with
+  | None -> test (List.hd paths) "didn't find ida!"
+  | Some ver ->
+    let x = List.find ~f:(fun p -> match version_of_path p with
```

Comment:
You can write without a pattern matching in a more concise way: `version_of_path p = Some ver`